### PR TITLE
dbwrapper: properly destroy objects in case CDBWrapper ctor throws

### LINKED ISSUE #22592: LeakSanitizer detects memory leak if CDBWrapper ctor fails
While running the functional tests under LSAN and 8 parallel jobs, `CDBWrapper` ctor would regularly throw via `HandleError` because of low disk space.  This would make LSAN error because the destructor for `CDBWrapper` was never called and so the `options.block_cache` and some other parameters wouldn't be destructed.

LSAN output (I deleted the file, but managed to take a picture of it, so log is kind of truncated):
```
==1204791==ERROR:  LeakSanitizer: detected memory leaks

Direct leak of 3520 byte(s) in 1 object(s) allocated from:
       #0 0x55f632912679 in operator new(unsigned long)
       #1 0x55f6342936d6 in leveldb::NewLRUCache(unsigned long)
       #2 0x55f6335f66f8 in GetOptions(unsigned long)
       #3 0x55f6335f4ec7 in CDBWrapper::CDBWrapper(boost::filesystem::path const&, unsigned long, bool, bool, bool)
       #4 0x55f6332e32cf in std::_MakeUniq<CDBWrapper>::__single_object std::make_unique<CoinsViews ...truncated...
       #5 0x55f6332d2781 in CCoinsViewDB::CCoinsViewDB(boost::filesystem::path, unsigned long, bool, bool)
       #6 0x55f6333ec6e2 in CoinsViews::CoinsViews(std::__cxx11::basic_string<char, std::char_traits<char>, ...truncated...
       #7 0x55f63346a111 in std::_MakeUniq<CoinsViews>::__single_object std::make_unique<CoinsViews, ...truncated...
       #8 0x55f6333ed379 in CChainState::InitCoinsDB(unsigned long, bool, bool, std::__cxx11::basic_string ...truncated...
       #9 0x55f6329699e0 in AppInitMain(NodeContext&, interfaces::BlockAndHeaderTipInfo*)
       #10 0x55f63291868e in AppInit(NodeContext&, int, char**)
       #11 0x55f632915e2f in main
       #12 0x7f86b6f7fb24 in __libc_start_main
```

### PR DESCRIPTION
**Pull request by Crypt-iQ**
There are two main changes: introduce LevelDbOptions to manage the leveldb::Options struct and changing CDBWrapper's penv member to a unique_ptr. The first allows LevelDbOptions to destroy leveldb::Options members without CDBWrapper having to directly. The second change allows the leveldb::Env to be managed by penv. Both of these changes prevent a LeakSanitizer exception if the CDBWrapper ctor throws.

Fixes https://github.com/bitcoin/bitcoin/issues/22592

### CODE REVIEW

**Comment by mrnguyenthanhtu** [comment]
Have a nice day!

**Comment by Crypt-iQ** [comment]
One build is failing due to a leak in `coins_view` and I cannot see why with the CI.  I am unable to reproduce this on my machine.

File: src/dbwrapper.cpp [hunk 1]
```diff
@@ -97,46 +97,59 @@ static void SetMaxOpenFiles(leveldb::Options *options) {
...
+            leveldb::Status result = leveldb::DestroyDB(path.string(), *options);
             dbwrapper_private::HandleError(result);
         }
         TryCreateDirectories(path);
         LogPrintf("Opening LevelDB in %s\n", path.string());
     }
-    leveldb::Status status = leveldb::DB::Open(options, path.string(), &pdb);
+    leveldb::Status status = leveldb::DB::Open(*options, path.string(), &pdb);
+    if (!status.ok()) {
+        delete penv;
```
Comment (Talkless) [inline src/dbwrapper.cpp:151]: If `penv` is deleted, then `~CDBWrapper()` gonna delete it again? Missing `= nullptr`?

Comment (Crypt-iQ) [reply src/dbwrapper.cpp:151]: If `!status.ok` then `HandleError` throws and the destructor isn't called

Comment (Talkless) [reply src/dbwrapper.cpp:151]: Maybe worth making `penv` a `unique_ptr` too, if we need manual handling with `delete`? Or do you think this would be kinda out of scope for this PR?

Adding addition `delete` into new code does not seem... "right" (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-newdelete).

Comment (Crypt-iQ) [reply src/dbwrapper.cpp:151]: good point, I can look into making it a `unique_ptr`

File: src/dbwrapper.cpp [hunk 2]
```diff
@@ -170,14 +183,7 @@ CDBWrapper::~CDBWrapper()
 {
     delete pdb;
```
Comment (ryanofsky) [inline src/dbwrapper.cpp:184]: In commit "dbwrapper: wrap options in unique_ptr and use DeleteOptions" (6770f066b69c6c271e7096cc647d5d42762f6b05)

I liked the suggestion to make `env` a unique_ptr https://github.com/bitcoin/bitcoin/pull/22663#discussion_r686783500, and maybe `pdb` could be a unique_ptr too?

File: src/dbwrapper.cpp [hunk 3]
```diff
@@ -97,46 +97,59 @@ static void SetMaxOpenFiles(leveldb::Options *options) {
...
-        options.paranoid_checks = true;
+        options->paranoid_checks = true;
     }
-    SetMaxOpenFiles(&options);
-    return options;
+    SetMaxOpenFiles(options);
+}
+
+// Custom deleter for unique_ptr<leveldb::Options>.
+static void DeleteOptions(leveldb::Options* options)
```
Comment (ryanofsky) [inline src/dbwrapper.cpp:116]: In commit "dbwrapper: wrap options in unique_ptr and use DeleteOptions" (6770f066b69c6c271e7096cc647d5d42762f6b05)

Custom deleter seems to be forgetting to `delete options`, which would explain the fuzzing leak? https://github.com/bitcoin/bitcoin/pull/22663#issuecomment-894792761

**Review by ryanofsky — APPROVED** [review]
6770f066b69c6c271e7096cc647d5d42762f6b05 looks pretty good if the leak is fixed. It generally seems to moves things in a good direction.

I do think more ideally, if the goal is to make the options struct more RAII-like and exception-safe, instead of requiring it to be allocated separately on the heap and requiring it to be paired with a custom deleter, it would be better to just inherit from it and write a normal destructor:

```c++
struct LevelDbOptions : public leveldb::Options {
  ~LevelDbOptions() {
    delete filter_policy;
    delete info_log;
    ...
  }
}
```

File: src/dbwrapper.cpp [hunk 4]
```diff
@@ -97,46 +97,59 @@ static void SetMaxOpenFiles(leveldb::Options *options) {
...
     if (leveldb::kMajorVersion > 1 || (leveldb::kMajorVersion == 1 && leveldb::kMinorVersion >= 16)) {
         // LevelDB versions before 1.16 consider short writes to be corruption. Only trigger error
         // on corruption in later versions.
-        options.paranoid_checks = true;
+        options->paranoid_checks = true;
     }
-    SetMaxOpenFiles(&options);
-    return options;
+    SetMaxOpenFiles(options);
+}
```
[the remark this answers is not in the document]
Comment (laanwj) [reply src/dbwrapper.cpp:113]: I agree with @ryanofsky. I don't think there's a strong need to change this function, but if there is, I would prefer it to return the target type `std::unique_ptr<leveldb::Options, void (*)(leveldb::Options*)` directly, instead of using a pointer out argument (which, strictly, seems a deterioration of API clarity to me).

[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply src/dbwrapper.cpp:113]: Will work on a patch with this, thanks for the review

**Comment by laanwj** [comment]
@Crypt-iQ can you please respond to @ryanofsky's comments and rebase ?

**Comment by Crypt-iQ** [comment]
Hi yes, I can get to it this weekend

File: src/dbwrapper.cpp (hunk 1 above)
Comment (Crypt-iQ) [reply src/dbwrapper.cpp]: The latest change makes `penv` a unique_ptr

File: src/dbwrapper.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply src/dbwrapper.cpp]: Latest change lifts the `GetOptions` logic into the ctor of `LevelDbOptions`

File: src/dbwrapper.cpp [hunk 5]
```diff
@@ -97,46 +97,59 @@ static void SetMaxOpenFiles(leveldb::Options *options) {
...
+// Custom deleter for unique_ptr<leveldb::Options>.
+static void DeleteOptions(leveldb::Options* options)
+{
+    delete options->filter_policy;
+    options->filter_policy = nullptr;
+    delete options->info_log;
+    options->info_log = nullptr;
+    delete options->block_cache;
+    options->block_cache = nullptr;
+    options->env = nullptr;
```
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply src/dbwrapper.cpp]: I opted for the second approach and dropped the line

File: src/dbwrapper.cpp (hunk 2 above)
Comment (Crypt-iQ) [reply src/dbwrapper.cpp:183]: I was not sure how to make `pdb` a unique_ptr and also pass it in to `Open` as `leveldb::DB**` when it is `nullptr` at this point. Maybe I am missing something.

File: src/dbwrapper.cpp (hunk 3 above)
Comment (Crypt-iQ) [reply src/dbwrapper.cpp]: No longer using a custom deleter and instead using `~LevelDbOptions`

**Comment by Crypt-iQ** [comment]
I've rebased and hopefully addressed all of the comments

File: src/dbwrapper.cpp [hunk 6]
```diff
@@ -97,36 +97,44 @@ static void SetMaxOpenFiles(leveldb::Options *options) {
...
 
-static leveldb::Options GetOptions(size_t nCacheSize)
-{
-    leveldb::Options options;
-    options.block_cache = leveldb::NewLRUCache(nCacheSize / 2);
-    options.write_buffer_size = nCacheSize / 4; // up to two write buffers may be held in memory simultaneously
-    options.filter_policy = leveldb::NewBloomFilterPolicy(10);
-    options.compression = leveldb::kNoCompression;
-    options.info_log = new CBitcoinLevelDBLogger();
+LevelDbOptions::LevelDbOptions() {
```
[the remark this answers is not in the document]
Comment (Crypt-iQ) [reply src/dbwrapper.cpp]: fixed

File: src/dbwrapper.h [hunk 7]
```diff
@@ -19,6 +19,17 @@
 static const size_t DBWRAPPER_PREALLOC_KEY_SIZE = 64;
 static const size_t DBWRAPPER_PREALLOC_VALUE_SIZE = 1024;
 
+struct LevelDbOptions : public leveldb::Options {
+    LevelDbOptions();
+
+    ~LevelDbOptions();
+
+    /** Sets block_cache and write_buffer_size since nCacheSize isn't known when
+     * LevelDbOptions ctor is called.
```
Comment (laanwj) [inline src/dbwrapper.h:28]: I don't think this is strictly true? Or am I missing something? `nCacheSize` is passed into `CDBWrapper::CDBWrapper`, couldn't you pass it in like
```
CDBWrapper::CDBWrapper(const fs::path& path, size_t nCacheSize, bool fMemory, bool fWipe, bool obfuscate)
    : m_name{fs::PathToString(path.stem())},
      options(nCacheSize)
⋮

struct LevelDbOptions : public leveldb::Options {
    LevelDbOptions(size_t nCacheSize);
```

**Review by laanwj — CHANGES_REQUESTED** [review]

File: src/dbwrapper.cpp (hunk 2 above)
Comment (ryanofsky) [reply src/dbwrapper.cpp:178]: re: https://github.com/bitcoin/bitcoin/pull/22663#discussion_r772488258

> I was not sure how to make `pdb` a unique_ptr and also pass it in to `Open` as `leveldb::DB**` when it is `nullptr` at this point. Maybe I am missing something.

No need to do this, but you would need to pass a temporary pointer to Open, and switch to unique_ptr after it returns:

```
leveldb::DB* new_db = nullptr;
leveldb::Status status = leveldb::DB::Open(options, fs::PathToString(path), &new_db);
pdb.reset(new_db);
```

**Review by ryanofsky — APPROVED** [review]
Code review ACK 2a5d230c0fd764004a8c83c022ba71d0ee95d2a3. This all looks good and is straightforward. Thanks for implementing all the suggested changes

File: src/dbwrapper.h [hunk 8]
```diff
@@ -178,10 +184,10 @@ class CDBWrapper
     friend const std::vector<unsigned char>& dbwrapper_private::GetObfuscateKey(const CDBWrapper &w);
 private:
     //! custom environment this database is using (may be nullptr in case of default environment)
-    leveldb::Env* penv;
+    std::unique_ptr<leveldb::Env> penv;
```
Comment (MarcoFalke) [inline src/dbwrapper.h:187]: nit: Maybe rename to m_env while touching all lines? Maybe split the penv change into a separate commit for easier review?

**Review by maflcko — APPROVED** [review]

File: src/dbwrapper.h (hunk 8 above)
Comment (Crypt-iQ) [reply src/dbwrapper.h]: renamed and split off into separate commit

Comment (Crypt-iQ) [reply src/dbwrapper.h]: I didn't functionally change the code and now there is a TSAN failure

Comment (MarcoFalke) [reply src/dbwrapper.h]: I couldn't make any sense of it, so reset it, but it is still here for anyone to look at: https://cirrus-ci.com/task/5612886578954240?logs=ci#L4869

Comment (Crypt-iQ) [reply src/dbwrapper.h]: I think this occurs and is unrelated to this PR because:
```
main thread:
* AppInitMain
  * ChainstateManager::ActiveTip()
    * CChain::Tip() -- accesses `vChain` without a lock (read)
```

```
b-loadblk thread:
* ThreadImport
  * ActivateBestChain called (src/node/blockstorage.cpp:883)
    * ActivateBestChainStep
      * ConnectTip
        * SetTip
          * vChain resize (write)
```

File: src/dbwrapper.h (hunk 7 above)
Comment (laanwj) [reply src/dbwrapper.h]: Thanks!

**Comment by achow101** [comment]
cc @jamesob

**Comment by Crypt-iQ** [comment]
unfortunately don't have time to work on this, lmk if i should close it

**Comment by fanquake** [comment]
> unfortunately don't have time to work on this, lmk if i should close it

Thanks. Closed and added up-for-grabs.
